clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3268,12 +3268,29 @@ static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &AL) {
 
   ASTContext &Ctx = S.getASTContext();
 
+  // The arguments to reqd_work_group_size are ordered based on which index
+  // increments the fastest. In OpenCL, the first argument is the index that
+  // increments the fastest, and in SYCL, the last argument is the index that
+  // increments the fastest.
+  //
+  // [[sycl::reqd_work_group_size]] and [[cl::reqd_work_group_size]] are
+  // available in SYCL modes and follow the SYCL rules.
+  // __attribute__((reqd_work_group_size)) is only available in OpenCL mode
+  // and follows the OpenCL rules.
   if (const auto *A = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
-    if (!((getExprValue(AL.getArgAsExpr(0), Ctx) <= *A->getXDimVal()) &&
-          (getExprValue(AL.getArgAsExpr(1), Ctx) <= *A->getYDimVal()) &&
-          (getExprValue(AL.getArgAsExpr(2), Ctx) <= *A->getZDimVal()))) {
+    bool CheckFirstArgument = S.getLangOpts().OpenCL
+              ? getExprValue(AL.getArgAsExpr(0), Ctx) <= *A->getZDimVal()
+	      : getExprValue(AL.getArgAsExpr(0), Ctx) <= *A->getXDimVal();
+    bool CheckSecondArgument =
+	      getExprValue(AL.getArgAsExpr(1), Ctx) <= *A->getYDimVal();
+    bool CheckThirdArgument = S.getLangOpts().OpenCL
+              ? getExprValue(AL.getArgAsExpr(2), Ctx) <= *A->getXDimVal()
+	      : getExprValue(AL.getArgAsExpr(2), Ctx) <= *A->getZDimVal();
+
+    if (!(CheckFirstArgument && CheckSecondArgument && CheckThirdArgument)) {
```

Comment:
@premanandrao, I tried this change in my ws before. It would break all existing tests

Reply:
@premanandrao, this would break all existing tests of reqd_work_group_size attribute implementation related to this diagnostic. I tried this before. I have verified again.

Reply:
Currently here we do:
```
p1 = a ? x <= y : x <= z;
p2 = ...
p3 = ...

if (!(p1 && p2 && p3))
```

I was suggesting we do:
```
p1 = a ? x > y: x > z;
p2 = ...
p3 = ...
then we do:
if (p1 || p2 || p3)
```

I believe these are equivalent (and matches the comparison form we use when calling checkWorkGroupSizeAttrValues()).   I am not sure why we get wrong diagnostics if we change this.

If it is not too much trouble, please point out what I am missing.   (I don't want to hold up the PR for this change, but I would like to understand it nonetheless.)

Reply:
The check would be `if (!CheckFirstArgument || !CheckSecondArgument || !CheckThirdArgument)`

Reply:
Oh wait. I didn't notice you changed what p1, p2 and p3 held here.

Reply:
Sorry for the confusion. Not sure what happened in my last testing (i probably made some mistake to change all augments correctly) .

This is fixed now.  

>>if (CheckFirstArgument || CheckSecondArgument || CheckThirdArgument) {

Done

Reply:
Thanks @smanna12.   Appreciate your changing this.   Hopefully a future refactoring can get this logic down to one place.  I agree that it is messy trying to get it right.

Reply:
>> Hopefully a future refactoring can get this logic down to one place. I agree that it is messy trying to get it right.

Currently reqd_work_group_size is pretty messy condition.  We have many existing bugs. I will start working on it after merging this PR. This part will be refactored i believe. Thanks @premanandrao for the suggeation.